src/main.rs

```diff
@@ -0,0 +1,270 @@
+        .header(header::ACCESS_CONTROL_ALLOW_HEADERS, header::CONTENT_TYPE)
+        .header(header::ACCESS_CONTROL_ALLOW_METHODS, "POST, OPTIONS")
+        .header(header::CONNECTION, "keep-alive")
+        .body(Body::from("No content"))?)
+}
+
+// Utility function to get a header value as a string. Returns an empty string
+// if no value is found.
+fn header_val(header: Option<&HeaderValue>) -> &str {
+    match header {
+        Some(h) => h.to_str().unwrap_or(""),
+        None => "",
+    }
+}
+
+/// Handle reports
+///
+/// It attempts to extract the NEL reports from the POST request body and maps
+/// over each report adding additional information before emitting a log line
+/// to the `reports` logging endpoint if valid. It always returns a synthetic
+/// `204 No content` response, regardless of whether the log reporting was
+/// successful.
+fn handle_reports(req: Request<Body>) -> Result<Response<Body>, Error> {
+    let (parts, body) = req.into_parts();
+
+    // Parse the NEL reports from the request JSON body using serde_json.
+    // If successful, bind the reports to the `reports` variable, transform and log.
+    if let Ok(reports) = serde_json::from_reader::<Body, Vec<Report>>(body) {
+        // Extract information about the client from the downstream request,
+        // such as the User-Agent and IP address.
+        let client_user_agent = header_val(parts.headers.get(header::USER_AGENT));
+        let client_ip = downstream_client_ip_addr().expect("should have client IP");
+
+        // Construct a new `ClientData` structure from the User-Agent and IP.
+        let client_data = ClientData::new(client_ip, client_user_agent)?;
+
+        // Map over each raw NEL report, merge it with the `ClientData` from
+        // above and transform it to a `LogLine` structure to merging it with the ClientData
+        let logs: Vec<LogLine> = reports
+            .into_iter()
+            .map(|report| LogLine::new(report, client_data.clone()))
+            .filter_map(Result::ok)
+            .collect();
+
+        // Create a handle to the upstream logging endpoint that we want to emit
+        // the reports too.
+        let mut endpoint = Endpoint::from_name("reports");
+
+        // Loop over each log line serializing it back to JSON and write it to
+        // the logging endpoint.
+        for log in logs.iter() {
+            if let Ok(json) = serde_json::to_string(&log) {
+                writeln!(endpoint, "{}", json)?;
+            }
+        }
+    };
+
+    // Return and empty 204 no content response to the downstream client,
+    // regardless of successful logging.
+    generate_no_content_response()
+}
+
+/// Main application entrypoint.
+///
+/// This controls the routing logic for the application, it accepts a `Request`
+/// and passes it to any matching request handlers before returning a `Response`
+/// back downstream.
+#[fastly::main]
+fn main(req: Request<Body>) -> Result<impl ResponseExt, Error> {
```

Comment:
Should `fn main()` and `fn handle_reports` come first in the file for a better flow?

Reply:
Haha funny our production version has the modules spilt out, but inlined here thinking it would be easier to follow. However I agree, and so does @triblondon, starter kits should be  examples of full applications. Have now refactored into  multiple files and moved `fn main` to the top. 👍